### PR TITLE
Modify translation map

### PR DESCRIPTION
**Pull request by knelli2**
## Proposed changes
Previously, each component of a translation had to be stored as a separate 1d translation map. If you wanted a 2d or 3d translation map, you'd need to do a product of maps. This simplifies the interface by templating the translation map on the dimension.

This was prompted by control system edits. For translation control, we only want one control system controlling one function of time that represents a translation in 3d. With the old implementation of the translation, we would have needed three separate control systems controlling three separate functions of time. That would not have been ideal.

### Upgrade instructions
- If you have any Translation maps here are the new versions:
```
Old  ->  New
Translation  ->  Translation<1>
ProductOf2Maps<Translation, Translation>  ->  Translation<2>
ProductOf3Maps<Translation, Translation, Translation>  ->  Translation<3>
```
- The UniformTranslation time dependence for a domain creator now takes as its last argument a `std::string` instead of a `std::array<std::string, Dim>` because we no longer need `Dim` number of functions of time to represent a translation map. We only need 1. Thus, in an input file, just a single string is needed for the name of the function of time, not a list.
- The option for the name of the function of time has been renamed from `FunctionOfTimeNames` to `FunctionOfTimeName`. Thus, with the above change, a typical UniformTranslation time dependence in 1 dimension would look like this in an input file:
```
UniformTranslation:
    InitialTime: 0.0
    InitialExpirationDeltaT: 2.0
    Velocity: [0.5]
    FunctionOfTimeName: "Translation1D"
```
and in 3 dimensions...
```
UniformTranslation:
    InitialTime: 0.0
    InitialExpirationDeltaT: 2.0
    Velocity: [0.5, 0.4, 0.3]
    FunctionOfTimeName: "Translation3D"
```

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: tests/Unit/Domain/Test_TagsTimeDependent.cpp [hunk 1]
```diff
@@ -128,11 +119,10 @@ void test() noexcept {
       domain::Tags::InertialMeshVelocityCompute<Dim>>;
 
   MAKE_GENERATOR(gen);
-  const std::array<double, 3> velocity{{1.2, 0.2, -8.9}};
+  const DataVector velocity{Dim, 1.2};
```
Comment (wthrowe) [inline tests/Unit/Domain/Test_TagsTimeDependent.cpp:120]: Why the change in value here?

File: tests/Unit/Evolution/Initialization/Test_DgDomain.cpp [hunk 2]
```diff
@@ -203,22 +176,19 @@ void test(const Spectral::Quadrature quadrature) noexcept {
   const std::vector<std::array<size_t, Dim>> initial_refinement{
       make_array<Dim>(0_st)};
   const size_t num_pts = pow<Dim>(4_st);
-  const std::array<double, 3> velocity{{1.2, 0.2, -8.9}};
+  const DataVector velocity{Dim, 3.6};
```
Comment (wthrowe) [inline tests/Unit/Evolution/Initialization/Test_DgDomain.cpp:177]: Again, why this change?

File: tests/Unit/Evolution/Test_TagsDomain.cpp [hunk 3]
```diff
@@ -114,14 +73,13 @@ void test() noexcept {
       domain::Tags::InertialMeshVelocityCompute<Dim>,
       evolution::domain::Tags::DivMeshVelocityCompute<Dim>>;
 
-  const std::array<double, 3> velocity{{1.2, 0.2, -8.9}};
+  const DataVector velocity{Dim, -4.3};
```
Comment (wthrowe) [inline tests/Unit/Evolution/Test_TagsDomain.cpp:74]: Again.

**Review by wthrowe — CHANGES_REQUESTED** [review]
In the future, if you are going to reformat a lot of code that you are not otherwise touching, please do it in a separate commit.

File: tests/Unit/Domain/Test_TagsTimeDependent.cpp (hunk 1 above)
Comment (knelli2) [reply tests/Unit/Domain/Test_TagsTimeDependent.cpp:120]: Below, on old lines 152-166, each component of the `std::array` was used to construct a `DataVector` of length 1 for each function of time (3 in total). Since we only need 1 function of time now, I just made the original `std::array` a `DataVector` of appropriate length (depending on the Dim).

Same for the other comments.

**Comment by knelli2** [comment]
Oops, guess I got a little carried away with the auto-formatting 😅 . I can split it up here if you want

File: tests/Unit/Domain/Test_TagsTimeDependent.cpp (hunk 1 above)
Comment (wthrowe) [reply tests/Unit/Domain/Test_TagsTimeDependent.cpp:120]: But why the change from {1.2, 0.2, -8.9} to {1.2, 1.2, 1.2}?

Comment (wthrowe) [reply tests/Unit/Domain/Test_TagsTimeDependent.cpp:120]: Oh, right, it needs to be the right size now.  Never mind.

Comment (knelli2) [reply tests/Unit/Domain/Test_TagsTimeDependent.cpp:120]: Oh the values are arbitrary so instead of adding `if` statements to construct a separate `DataVector` for each Dim I just did this since it was simpler.

**Review by wthrowe — DISMISSED** [review]
Up to you on splitting the commit.  This is long enough that I want another reviewer, and they may ask for it to be done, though.

**Comment by knelli2** [comment]
Ok I decided to just undo the formatting changes that weren't relevant to this PR. I think I got everything. Let me know if I didn't.

**Review by wthrowe — DISMISSED** [review]

File: src/Domain/CoordinateMaps/TimeDependent/Translation.cpp [hunk 4]
```diff
@@ -62,58 +87,94 @@ std::array<tt::remove_cvref_wrap_t<T>, 1> Translation::frame_velocity(
                                   << "' is not one of the known functions of "
                                      "time. The known functions of time are: "
                                   << keys_of(functions_of_time));
-  return {{make_with_value<tt::remove_cvref_wrap_t<T>>(
-      dereference_wrapper(source_coords[0]),
-      functions_of_time.at(f_of_t_name_)->func_and_deriv(time)[1][0])}};
+  std::array<tt::remove_cvref_wrap_t<T>, Dim> result{};
+  const auto function_of_time_and_deriv =
+      functions_of_time.at(f_of_t_name_)->func_and_deriv(time);
+  const DataVector velocity = function_of_time_and_deriv[1];
```
Comment (nilsdeppe) [inline src/Domain/CoordinateMaps/TimeDependent/Translation.cpp]: I thought I explicitly asked when this was discussed if the velocity would be `std::array<double, Dim>` and not a `DataVector` to which I remember being told "yes". I guess this isn't true? Am I misremembering what I was told could be a `std::array<double>`? I don't have a change request, I'm just confused that this disagrees with what I remember. I guess previously it was a `DataVector` as well instead of a `double` or `std::arary<double, 1>`?

`const DataVector` -> `const DataVector&` should work, no?

File: tests/Unit/Domain/Creators/TimeDependence/Test_UniformTranslation.cpp [hunk 5]
```diff
@@ -31,54 +31,24 @@
 namespace domain::creators::time_dependence {
 
 namespace {
```
Comment (nilsdeppe) [inline tests/Unit/Domain/Creators/TimeDependence/Test_UniformTranslation.cpp:31]: is now possible to remove the `ProductMaps.?pp` includes?

File: tests/Unit/Domain/Test_Block.cpp [hunk 6]
```diff
@@ -99,52 +99,19 @@ void test_block_time_independent() {
   test_move_semantics(std::move(original_block), block_copy);
 }
```
Comment (nilsdeppe) [inline tests/Unit/Domain/Test_Block.cpp:99]: do you need the time-dependent product maps include still?

File: tests/Unit/Domain/Test_TagsTimeDependent.cpp [hunk 7]
```diff
@@ -58,13 +58,9 @@ void test_tags() noexcept {
       "div(MeshVelocity)");
 }
```
Comment (nilsdeppe) [inline tests/Unit/Domain/Test_TagsTimeDependent.cpp:58]: remove include of time-dependent product maps?

File: tests/Unit/Evolution/Initialization/Test_DgDomain.cpp [hunk 8]
```diff
@@ -39,12 +39,9 @@
 #include "Utilities/TMPL.hpp"
 
 namespace {
```
Comment (nilsdeppe) [inline tests/Unit/Evolution/Initialization/Test_DgDomain.cpp:39]: remove include of time-dependent product maps?

File: tests/Unit/Evolution/Test_TagsDomain.cpp [hunk 9]
```diff
@@ -43,57 +42,17 @@ void test_tags() noexcept {
       "div(MeshVelocity)");
 }
```
Comment (nilsdeppe) [inline tests/Unit/Evolution/Test_TagsDomain.cpp:42]: remove include of time-dependent product maps?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
Overall looks good. Just a few comments. Please squash directly :)

File: src/Domain/CoordinateMaps/TimeDependent/Translation.cpp (hunk 4 above)
Comment (knelli2) [reply src/Domain/CoordinateMaps/TimeDependent/Translation.cpp]: Hm I guess I'm confused by your confusion. Nothing about the outward interface of this function has changed. The source coordinates input type is still either a `std::array<double>` or `std::array<DataVector>` and the return type is still a `std::array<double>` or `std::array<DataVector>` corresponding to the input type (see line 90). Nothing has changed in that regard.

The velocity defined on line 93 is just a temporary. The `make_with_value` on line 100 constructs the appropriate return type (`double` or `DataVector`) for each element of the resulting `std::array`. Does that answer your question?

And yes, `const DataVector&` works, I'll put that in.

File: tests/Unit/Domain/Creators/TimeDependence/Test_UniformTranslation.cpp (hunk 5 above)
Comment (knelli2) [reply tests/Unit/Domain/Creators/TimeDependence/Test_UniformTranslation.cpp:31]: Oops, yup forgot to take out all these includes.

File: src/Domain/CoordinateMaps/TimeDependent/Translation.cpp (hunk 4 above)
Comment (nilsdeppe) [reply src/Domain/CoordinateMaps/TimeDependent/Translation.cpp]: Yea, I was probably just confused at the time of the discussion I'm remembering :) The code here is what I expect based on what I wrote 😄 

:+1:

**Review by nilsdeppe — APPROVED** [review]
@wthrowe I'm happy for this to be merged

**Review by wthrowe — APPROVED** [review]
